crates/pathfinder/src/rpc/v02/method/get_proof.rs

```diff
@@ -0,0 +1,213 @@
+use stark_hash::StarkHash;
+
+#[derive(Deserialize, Debug, PartialEq, Eq)]
+pub struct GetProofInput {
+    pub contract_address: ContractAddress,
+    pub keys: Vec<StorageAddress>,
+    pub block_id: BlockId,
+}
+
+crate::rpc::error::generate_rpc_error_subset!(GetProofError: BlockNotFound);
+
+/// Utility struct used for serializing.
+#[derive(Debug, Serialize)]
+struct PathWrapper {
+    value: StarkHash,
+    len: usize,
+}
+
+impl Serialize for ProofNode {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        match &self {
+            ProofNode::Binary(bin) => {
+                let mut state = serializer.serialize_struct_variant("ProofNode", 0, "Binary", 2)?;
+                state.serialize_field("left", &bin.left_hash)?;
+                state.serialize_field("right", &bin.right_hash)?;
+                state.end()
+            }
+            ProofNode::Edge(edge) => {
+                let value = StarkHash::from_bits(edge.path.as_bitslice()).unwrap();
+                let path_wrapper = PathWrapper {
+                    value,
+                    len: edge.path.len(),
+                };
+
+                let mut state = serializer.serialize_struct_variant("ProofNode", 1, "Edge", 2)?;
+                state.serialize_field("path", &path_wrapper)?;
+                state.serialize_field("child", &edge.child_hash)?;
+                state.end()
+            }
+        }
+    }
+}
+
+/// Holds the data and proofs for a specific contract.
+#[derive(Debug, Serialize)]
+pub struct ContractData {
+    // Required to verify the contract state hash to contract root calculation.
+    class_hash: ClassHash,
+    // Required to verify the contract state hash to contract root calculation.
+    nonce: ContractNonce,
+
+    // Root of the Contract state tree
+    root: ContractRoot,
+
+    // This is currently just a constant = 0, however it might change in the future.
+    contract_state_hash_version: StarkHash,
+
+    // The proofs associated with the queried storage values
+    storage_proofs: Vec<Vec<ProofNode>>,
+}
+
+/// Holds the membership/non-membership of a contract and its associated contract contract if the contract exists.
+#[derive(Debug, Serialize)]
+pub struct GetProofOutput {
+    // Membership / Non-membership proof for the queried contract
+    contract_proof: Vec<ProofNode>,
+
+    // Additional contract data if it exists.
+    contract_data: Option<ContractData>,
+}
+
+/// Returns all the necessary data to trustlessly verify storage slots for a particular contract.
+pub async fn get_proof(
+    context: RpcContext,
+    input: GetProofInput,
+) -> Result<GetProofOutput, GetProofError> {
+    let block_id = match input.block_id {
+        BlockId::Hash(hash) => hash.into(),
+        BlockId::Number(number) => number.into(),
+        BlockId::Latest => StarknetBlocksBlockId::Latest,
+        BlockId::Pending => {
```

Comment:
If pending is generally not supported by this method then imo the fallback to latest should be removed and an internal error should always be returned here.

Reply:
eefeec3b20d03ec16fc2f185a9085ec1739f601d